core-primitives/node-api-extensions/src/metadata.rs

```diff
@@ -0,0 +1,192 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+//! Handle SGX compatible substrate chain metadata.
+
+use crate::error::{Error, Result};
+use codec::{Decode, Encode};
+use sp_core::storage::StorageKey;
+use substrate_api_client::{Metadata, MetadataError};
+
+/// Pallet's names:
+static TEEREX: &str = "Teerex";
+static SIDECHAIN: &str = "Sidechain";
+static SYSTEM: &str = "System";
+
+#[derive(Default, Encode, Decode, Debug, Clone)]
+pub struct NodeMetadata {
+	node_metadata: Option<Metadata>,
+	runtime_spec_version: u32,
+	runtime_transaction_version: u32,
+}
+
+impl NodeMetadata {
+	pub fn new(
+		node_metadata: Metadata,
+		runtime_spec_version: u32,
+		runtime_transaction_version: u32,
+	) -> Self {
+		Self {
+			node_metadata: Some(node_metadata),
+			runtime_spec_version,
+			runtime_transaction_version,
+		}
+	}
+
+	/// Teerex storages:
+	pub fn enclave_count_storage_key(&self) -> Result<StorageKey> {
+		self.storage_value_key(TEEREX, "EnclaveCount")
+	}
+
+	pub fn enclave_registry_storage_map_key(&self, index: u64) -> Result<StorageKey> {
+		self.storage_map_key(TEEREX, "EnclaveRegistry", &index)
+	}
+
+	/// System storages:
+	pub fn system_account_storage_key(&self) -> Result<StorageKey> {
+		self.storage_value_key(SYSTEM, "Account")
+	}
+
+	pub fn system_account_storage_map_key(&self, index: u64) -> Result<StorageKey> {
+		self.storage_map_key(SYSTEM, "Account", &index)
+	}
```

Comment:
I would prefer to separate these things:

* only implement `storage_value_key`, `storage_map_key`, and `storage_double_map_key` on `NodeMetadata` directly.
* The rest should be in different traits: `Teerex`, `System`

Organized in a module:

* metadata/mod.rs // contains the metadata definition
* metadata/teerex.rs
* metadata/system.rs

Reply:
This will also stimulate good design style downstream, to implement their own traits in separate files crates, which will reduce merge conflicts, if they want to extend the available cals here.

Reply:
So the grouping currently is by functionality: `CallIndexes`, `RuntimeVersion` and `StorageKeys`. Since we don't use the storage key functionality yet, we left them here in the `impl`. This should be changed once we make the next step and actually use these storage keys, agreed.

What you're suggesting is a bit of a different grouping, I believe? You want all Teerex related metadata in a single trait? That would include the call indexes AND storage keys?

Reply:
I agree with the `RuntimeVersion` and generic `StorageKeys` in the main `impl` blocks, but I'd add a grouping for the pallet-specific stuff. I think the nicest separation would be a separate `TeeracleCalls` and `TeeracleStorageKey` trait in a teeracle module.
What do you think about that?

Reply:
When you say `Teeracle`, you mean `Teerex`? So we'd have a `Teerex`, a `Sidechain` and `System` module, with traits for the call indexes and the storage keys for each (except `System`, which has no call indexes)?.